File: python-leetcode/linked-list/21.合并两个有序链表.py
from typing import Optional

from type import ListNode


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        # 边界情况：两个链表都为空
        if not list1 and not list2:
            return None

        dummy = ListNode(-1)
        curr = dummy

        while list1 and list2:
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next

        # 连接剩余部分
        curr.next = list1 or list2

        return dummy.next

File: go-leetcode/linked-list/21.合并两个有序链表.go
package leetcode

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// 边界情况：两个链表都为空
	if list1 == nil && list2 == nil {
		return nil
	}

	dummy := &ListNode{Val: -1}
	curr := dummy

	for list1 != nil && list2 != nil {
		if list1.Val <= list2.Val {
			curr.Next = list1
			list1 = list1.Next
		} else {
			curr.Next = list2
			list2 = list2.Next
		}
		curr = curr.Next
	}

	// 连接剩余部分
	if list1 != nil {
		curr.Next = list1
	} else {
		curr.Next = list2
	}

	return dummy.Next
}
